# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGL2ContextRenderbuffers.cpp
# Commit: 71c122ac0ca7
# Full Hash: 71c122ac0ca73391866b1ef19f4f82bc2d28568b
# Author: David Parks <davidp99@gmail.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Initial out-of-process WebGL implementation. r=mccr8,handyman
#   
#   Splits WebGLContext into ClientWebGLContext and HostWebGLContext.  The Client enables the JS-control of a WebGL context in a content procecss while the Host executes the WebGL graphics operations (via a WebGLContext that maintains much of the existing code) in the compositor process.  At this point, the cross-process behavior is disabled -- this series of patches is an incremental step toward that final goal.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D54018
# ==============================================================================

diff -r 5cbf5bb11c58 -r 71c122ac0ca7 dom/canvas/WebGL2ContextRenderbuffers.cpp
--- a/dom/canvas/WebGL2ContextRenderbuffers.cpp	Wed Jan 08 21:52:03 2020 +0000
+++ b/dom/canvas/WebGL2ContextRenderbuffers.cpp	Wed Jan 08 22:19:14 2020 +0000
@@ -12,19 +12,14 @@
 
 namespace mozilla {
 
-void WebGL2Context::GetInternalformatParameter(JSContext* cx, GLenum target,
-                                               GLenum internalformat,
-                                               GLenum pname,
-                                               JS::MutableHandleValue retval,
-                                               ErrorResult& out_rv) {
+Maybe<nsTArray<int32_t>> WebGL2Context::GetInternalformatParameter(
+    GLenum target, GLenum internalformat, GLenum pname) {
   const FuncScope funcScope(*this, "getInternalfomratParameter");
-  retval.setObjectOrNull(nullptr);
-
-  if (IsContextLost()) return;
+  if (IsContextLost()) return Nothing();
 
   if (target != LOCAL_GL_RENDERBUFFER) {
     ErrorInvalidEnum("`target` must be RENDERBUFFER.");
-    return;
+    return Nothing();
   }
 
   // GLES 3.0.4 $4.4.4 p212:
@@ -55,31 +50,31 @@
         "`internalformat` must be color-, depth-, or stencil-renderable, was: "
         "0x%04x.",
         internalformat);
-    return;
+    return Nothing();
   }
 
   if (pname != LOCAL_GL_SAMPLES) {
     ErrorInvalidEnum("`pname` must be SAMPLES.");
-    return;
-  }
-  std::vector<GLint> samples;
-  const auto maxSamples = usage->MaxSamples(*gl);
-  if (maxSamples) {  // It might be force-set to 0 for validation reasons!
-    GLint sampleCount = 0;
-    gl->fGetInternalformativ(LOCAL_GL_RENDERBUFFER, internalformat,
-                             LOCAL_GL_NUM_SAMPLE_COUNTS, 1, &sampleCount);
-    samples.resize(uint32_t(sampleCount));
-    gl->fGetInternalformativ(LOCAL_GL_RENDERBUFFER, internalformat,
-                             LOCAL_GL_SAMPLES, samples.size(), samples.data());
+    return Nothing();
   }
 
-  JSObject* obj =
-      dom::Int32Array::Create(cx, this, samples.size(), samples.data());
-  if (!obj) {
-    out_rv = NS_ERROR_OUT_OF_MEMORY;
+  nsTArray<int32_t> obj;
+  GLint sampleCount = 0;
+  gl->fGetInternalformativ(LOCAL_GL_RENDERBUFFER, internalformat,
+                           LOCAL_GL_NUM_SAMPLE_COUNTS, 1, &sampleCount);
+  if (sampleCount > 0) {
+    GLint* samples =
+        static_cast<GLint*>(obj.AppendElements(sampleCount, fallible));
+    // If we don't have 'samples' then we will return a zero-length array, which
+    // will be interpreted by the ClientWebGLContext as an out-of-memory
+    // condition.
+    if (samples) {
+      gl->fGetInternalformativ(LOCAL_GL_RENDERBUFFER, internalformat,
+                               LOCAL_GL_SAMPLES, sampleCount, samples);
+    }
   }
 
-  retval.setObjectOrNull(obj);
+  return Some(obj);
 }
 
 }  // namespace mozilla